[WIP] Grants

KennethAshley opened the pull request:
##### Description
This aims to add Grants to Gitcoin

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] linter status: 100% pass
- [ ] changes don't break existing behavior
- [ ] commit message follows [commit guidelines](https://github.com/gitcoinco/web/blob/master/docs/CONTRIBUTING.md#step-4-commit)

##### Affected core subsystem(s)
BE, UI, FE

##### Note
I know my way around the BE but i'm not master. Would love some more eyes on my BE code.

##### Refers/Fixes
Adds #380 

#### Todo
- [x] Add create grants from
- [ ] Allow image upload
- [ ] Allow adding of stake holders
- [x] form submission
- [x] Calculate percentage done between current funding and goal funding
- [x] List out Grants
- [x] Add progress bar
- [x] Style grants on index page
- [x] Style show page
- [ ] Add "fund project" interaction
- [ ] Add Grants initiative page

owocki commented:
very excited for this!  @ me back when its ready for QA

owocki commented:
i have a call with the founder of staketree.com soon to discuss maybe using some of their tech / product ideas here

KennethAshley commented:
@owocki @PixelantDesign added the html and some js here is the view

![screencapture-localhost-8000-grants-new-2018-03-29-21_01_05](https://user-images.githubusercontent.com/1138619/38120114-8165314a-3394-11e8-852f-7c6f338c3f59.png)

KennethAshley commented:
@owocki @mbeacom  I have added the following:

- listing grants on the `grants/explorer' page.
- adding `percentage_done` method
- added profile to `Grant` class (filled in from current user)
- styled elements

here is what it currently looks like:

![screencapture-localhost-8000-grants-explorer-2018-04-01-18_52_03](https://user-images.githubusercontent.com/1138619/38178351-002d2c02-35df-11e8-8ae6-dfdeaea941b6.png)

 @PixelantDesign is working on the top area

KennethAshley commented:
@owocki @PixelantDesign coded up the show page. I gotta work on the interaction for when a user presses "fund this project". 

![screencapture-localhost-8000-grant-14-2018-04-02-16_17_22](https://user-images.githubusercontent.com/1138619/38214044-628ae652-3691-11e8-90d7-0b8076574939.png)

owocki commented:
i'm ont convinced we need an extra landing page for this.. lets fold `grants/explorer` into `/grants/` and maybe just add some UI elements at the top explaining what the grants are

owocki commented:
looks like the model migrations`0046_auto_20180318_1054.py` and `0048_auto_20180320_1719.py` both create Grants.  the migrations on this PR need to be rolled back and recreated i think..

owocki commented:
858c708602fdd386a0836d62247c452789236e05 should cleanup the grants

owocki commented:
looks like we need to handle the case in which theres no logged in profile at `grants/new` http://bits.owocki.com/2T1t3m1P0g15/Screen%20Shot%202018-04-05%20at%204.03.04%20PM.png

owocki commented:
these form elements on grants/new need cleanup:
http://bits.owocki.com/1q0T032y2U0K/Screen%20Shot%202018-04-05%20at%204.05.49%20PM.png

drag & drop doesnt work
how much do you need 'doesnt make sense' or is not labeled correctly
'who is involved in this project should allow multiple email addresses

owocki commented:
grants explorer page needs some TLC around its CSS http://bits.owocki.com/1H0I2S0z0N0I/Screen%20Shot%202018-04-05%20at%204.06.49%20PM.png

owocki commented:
i believe we'll also need to curate the grants we show at `grant/explorer` to prevent anyone from just submitting a grant with some offensive content in it.  the system should probably send an email to the admin to list it on gitcoin.co/grants/exploer

we could probably follow the pattern at http://localhost:8080/universe/new (`accepted` flag.. email to admin, etc)

owocki commented:
grants/show/ page needs some TLC also... all of the profile info is just stubbed out over there

owocki commented:
just spent about an hour digging into https://github.com/StakeTree/contracts as the smart contract layer for this

its definitely technically feasible.. though the scope is much larger than i thought it was before i started my notes => http://bits.owocki.com/191U0V2a3V2V/Archive%202018-04-06%20at%201.58.50%20PM.zip

mbeacom commented on app/dashboard/migrations/0047_auto_20180320_1718.py:
```diff
@@ -0,0 +1,23 @@
+# Generated by Django 2.0.3 on 2018-03-21 00:18
```
You'll need to delete and recreate your added migrations for `dashboard` in this branch.

KennethAshley replied:
was this something that @owocki did above?

mbeacom commented on app/dashboard/views.py:
```diff
@@ -430,6 +430,56 @@ def dashboard(request):
+
+def grants(request):
+    """Handle grants explorer."""
+    grants = Grant.objects.all();
```
You'll want to drop the `;` here.

Also, should we consider pagination of some kind for grants?

KennethAshley replied:
whoops thats my js knowledge coming in.

Pagination would be a good idea.

mbeacom commented on app/dashboard/views.py:
```diff
@@ -430,6 +430,56 @@ def dashboard(request):
+        'active': 'dashboard',
+        'title': 'Grants Explorer',
+        'grants': grants,
+        'keywords': json.dumps([str(key) for key in Keyword.objects.all().values_list('keyword', flat=True)]),
```
Nothing really you're doing here, but this makes me think we should migrate to something like `taggit` for keywords.

mbeacom commented on app/app/urls.py:
```diff
@@ -49,6 +49,11 @@
+    # grant views
+    path('grant/<int:grant_id>/', dashboard.views.grant_show, name='grant'),
+    url(r'^grants/new?', dashboard.views.new_grant, name='new_grant'),
+    url(r'^grants/?', dashboard.views.grants, name='grants'),
```
Can you modify these two `url` entries to the new `re_path`? Should be a matter of just replacing `url(` with `re_path(`

KennethAshley commented:
@owocki @PixelantDesign i added intro content to the grants page

![screencapture-localhost-8000-grants-explorer-2018-04-11-21_04_12](https://user-images.githubusercontent.com/1138619/38650886-59f5ed10-3dcc-11e8-939f-7e5fe2feb44e.png)

thelostone-mc commented on app/retail/templates/shared/head.html:
```diff
@@ -39,13 +39,16 @@
     <link rel="stylesheet" href="{% static "v2/css/forms/select.css" %}" />
     <link rel="stylesheet" href="{% static "v2/css/external_bounties/checkboxes.css" %}" />
 
+    <link rel="stylesheet" href="{% static "v2/css/grants/grant.css" %}" />
```
@KennethAshley wanna move this into grant pages, rather than keeping in in the header?

mbeacom commented:
@KennethAshley Anything I can do to help you finish this up?
Thanks for the contribution!

owocki commented:
i believe i'm the blocker here.. i owe @KennethAshley the integration of staketree smart contracts here.. (unless we want to launch lean and mean and just do the feature off-chain completely)

eayoungs commented:
RE:
    _Quick question: Looks like about ~200 / 315 test are being skipped from the first commit. Do you 
     have any insight into why that its?_

@KennethAshley After running pytest with the -rs flag, it looks like the skipped tests are due to isort.
`SKIP [200] /usr/local/lib/python3.6/site-packages/pytest_isort.py:152: file(s) previously passed isort checks`

eayoungs commented:
First thing I ran into was.
`CommandError: Conflicting migrations detected; multiple leaf nodes in the migration graph: (0057_auto_20180419_0910, 0051_grant_stakeholder in dashboard).
To fix them run 'python manage.py makemigrations --merge'
I assume this is what @mbeacon was referring to above.

Once the merge was made, I had _4_ failing tests:
`CommandError: Conflicting migrations detected; multiple leaf nodes in the migration graph: (0057_auto_20180419_0910, 0051_grant_stakeholder in dashboard).
To fix them run 'python manage.py makemigrations --merge'`

mbeacom commented:
Updated this PR branch, resolved conflicts, and fixed the migrations.

eswarasai commented:
@owocki @captnseagraves -- Can we close this PR in the view of having `grants` being tracked under `web:grants` branch?

cc: @octavioamu

thelostone-mc commented:
@eswarasai is this good to close?

eswarasai commented:
@thelostone-mc -- Yep. Was about to do that. Will create a separate PR for `web:grants` to merge into `web:master` to keep track of it.
